Add package cache to support module resolution in analytics pipelines

sadhansood opened the pull request:
## Description 

Analytics pipeline v2 needs to support move objects in json format for querying (to bring parity with older pipeline) which requires access to package modules. This PR adds a persisted cache locally as a stop gap solution for package resolution until data processing framework provides a better alternative.

## Test Plan 

Ran the pipeline locally to ensure cache is working. The deployment will require some manual work to first build the local persisted cache from previous checkpoints.

sadhansood commented on crates/sui-analytics-indexer/src/handlers/mod.rs:
```diff
@@ -153,3 +156,20 @@ impl ObjectStatusTracker {
+    contents: &[u8],
+    package_cache: &PackageCache,
+) -> Result<MoveStruct> {
+    let move_struct = match package_cache
```
@amnn would this be the right way to convert a sui object to `MoveStruct` while using package cache?

amnn replied:
This looks right to me! (But of course, the real measure of correctness is whether tests that involve both types defined in original and updated packages pass).

sadhansood commented on crates/sui-package-resolver/src/store.rs:
```diff
@@ -0,0 +1,225 @@
+    async fn fetch(&self, id: AccountAddress) -> Result<Package>;
+
+    /// Store package `object` in the underlying store
+    async fn update(&self, object: &Object) -> Result<()>;
```
this function was added to the existing trait to support underlying store with new packages

sadhansood commented on crates/sui-package-resolver/src/store.rs:
```diff
@@ -0,0 +1,225 @@
+/// kept updated with latest version of package objects while iterating over checkpoints. If the
+/// local db is missing (or gets deleted), packages are fetched from a full node and local store is
+/// updated
+pub struct LocalDBPackageStore {
```
This struct implements `PackageStore` trait by keeping a local rocksdb store of package objects. DB is kept updated with latest package versions while iterating over checkpoints in pipelines

bmwill commented on crates/sui-package-resolver/src/store.rs:
```diff
@@ -0,0 +1,225 @@
+    async fn update(&self, object: &Object) -> Result<()>;
+}
+
+pub struct DbPackageStore(pub(crate) IndexerReader);
```
In order to have a better separation of dependencies I think that we shouldn't have any implementations of "PackageStore" (except for maybe a purely in-memory one) live in this crate. Instead we should export the trait and then let downstream consumers build and implement their own package caches using the dependencies that they want to bring in.

As it stands in this PR right now, in order to use this package resolver you need to depend on rocksdb and postgres despite the analytics pipeline not needing to depend on postgres and the graphql servers not needing to depend on rocksdb.

Alternatively we can leverage feature flags to bring those implementations and dependencies...

I mean today we don't really worry much about these things so we could punt on this but eventually we as a project need to do a bit better with how we're structuring these things. 

@amnn What are your thoughts on this?

amnn replied:
> Alternatively we can leverage feature flags

Please, no 😭 .

bmwill replied:
The setup that seems to work best is to:
- Have a `core` base crate which defines the interface and maybe one simple, no dependency, implementation
- All implementations exist in their own small crate or live in the crate of where they're consumed/used
- You can have a mega "sdk" like crate that uses feature flags to enable the implementations that the user wants by optionally enabling the dependencies where those implementations exist, and rexporting them. this gives you a single crate that users can depend on.

That last one is not very good for our own usecases internally but could be interesting for external user. My recommendation is we do the first two bullets and ignore the third for now. Which seems to be inline with your preferences too.

bmwill commented on crates/sui-package-resolver/src/store.rs:
```diff
@@ -0,0 +1,225 @@
+/// Interface to abstract over access to a store of live packages.  Used to override the default
+/// store during testing.
+#[async_trait]
+pub trait PackageStore {
```
@amnn I think we may want to discuss retooling this trait and the coupled Cache. As it stands today there are a few issues I see with it.

- The `version` function is only used to check for newer versions of the system packages since they are the only packages that can increase in version number. This event of them increasing in version number is a very, very rare occurrence and yet we're currently paying a rather large penalty, a single round-trip, for every single occurrence of a system package type when resolving a type. Given the framework types are far more likely to show up than any others this seems like a bad trade-off for a cache. I'd propose we remove this mechanism entirely and instead provide a method on  the cache for forcefully evicting packages from the cache so that policies for how/when we fallback to the refresh system packages can be run outside of the critical path of doing a type resolution.
- @sadhansood The addition of the `update` function seems out of place for what i perceive to be the intent behind this trait which is that the backing package store is to be used as a read-only store and not something that you can update. Instead I think that this `update` function seems like a prime candidate to not be apart of this trait but instead a method on the `PackageCache` itself used to update the local cache state, preventing the need to reach out to some remote store for the update. If you do indeed want to have a way to update the backing package store, in the case of having a rocksdb local store, i'd hold onto a handle to the store and update it directly there.

amnn replied:
Looks like we raced on reviews!

> The version function is only used to check for newer versions of the system packages since they are the only packages that can increase in version number.

> I'd propose we remove this mechanism entirely and instead provide a method on the cache for forcefully evicting packages from the cache so that policies for how/when we fallback to the refresh system packages can be run outside of the critical path of doing a type resolution.

A system to forcefully evict packages works well in the data pipeline cases where data is being pushed through the pipeline, but it does not work so well for the GraphQL service which is pulling data, so has no idea when to evict.  Any issues with the plan we discussed in Denver which was to put a type layout cache in front of the package cache.?  Framework types, which show up often, will remain in this cache.  Therefore we won't hit the slow path with the version check in it often, or at all in practice, unless somebody queries a layout involving a brand new framework type.

Also open to suggestions that would allow the pipelines to evict on push, but the GraphQL service to re-query on pull.  In another comment on this PR, I suggested fully encapsulating the storage interface (i.e. no `LruCache` by default) maybe that gets us what we want along this vein?

> The addition of the update function seems out of place for what i perceive to be the intent behind this trait which is that the backing package store is to be used as a read-only store and not something that you can update.

I tend to agree here as well -- this was another thing that made me want to change how the store interface worked here (although, for what it's worth, I was also planning on doing something similar to this in the GraphQL service, for the case where a package is loaded as an object first, and then queried for its structs and functions -- it's useful in that case to be able to inject a package into the "store", but in the GraphQL case the main benefit is to update the `LruCache`, not touch the backing store).

bmwill replied:
> but it does not work so well for the GraphQL service which is pulling data, so has no idea when to evict

We had a discussion a bit ago about how subscriptions would be handled in the GraphQL service, specifically that one approach would be for the GraphQL service to poll the DB for updates. If we go that route then it seems like we could also leverage that same mechanism for when to evict system packages

amnn replied:
The difference for me is that the push-pull mechanism for subscriptions means that for downstream clients, it is the "leading edge" that causes them to query further things, and all we need to do is make sure that before we issue the subscription update, everything else in the store is consistent (and polling is fine/good for this).  If we use the same mechanism for updating the underlying store (in this cache the type layout "store"), a query could fail because we didn't poll for the system package update fast enough, although everything else is otherwise consistent.

amnn commented on crates/sui-package-resolver/Cargo.toml:
```diff
@@ -0,0 +1,41 @@
+diesel.workspace = true
+move-binary-format.workspace = true
+move-core-types.workspace = true
+rocksdb.workspace = true
```
I'd like to avoid taking this dependency for the GraphQL service.  Can we split out the implementations of the various stores so that the analytics pipeline implements the local db store and has the `rocksdb` dependency, and GraphQL implements the existing `DbPackageStore` based on postgres/diesel?

amnn commented on crates/sui-package-resolver/src/error.rs:
```diff
@@ -0,0 +1,59 @@
+pub enum Error {
+    #[error("{0}")]
+    Bcs(#[from] bcs::Error),
+
+    #[error("{0}")]
+    Db(#[from] DbError),
+
+    #[error("{0}")]
+    TypedStore(#[from] TypedStoreError),
```
Now that we have multiple stores under the hood, we will likely need to make this error type generic, with an underlying `StoreError`:

```rust
pub enum Error<StoreError: Error> {
    #[error(transparent)]
    Store(#[from] StoreError),
}
```

bmwill replied:
The better approach would likely be to use a trait object here:
```
pub enum Error {
    Store(Box<dyn Error + Send + 'static>),
}
```

imo theres little value in having generic error types and it makes things harder to work with. If you do need to recover the concrete type you can leverage downcasting

amnn replied:
That's fine as well.

bmwill commented on crates/sui-package-resolver/src/cache.rs:
```diff
@@ -0,0 +1,96 @@
+/// object and parsing its contents.
+pub struct PackageCache {
+    packages: Mutex<LruCache<AccountAddress, Arc<Package>>>,
+    store: Box<dyn PackageStore + Send + Sync>,
```
@amnn Was there a particular reason for making this a trait object vs a generic?

amnn replied:
I can't think of a reason, and I'm not opposed to this being made generic -- I don't think I tried that in this case.

bmwill replied:
I don't think we need to change it, I was just asking because if this was generic we could add functions to get a reference to the internal store so that different users could call type specific methods on the internal store, when the concrete type is know in a context. This could prevent the need to hold onto a second handle to the rocksdb database to do updates in the case of the analytics pipeline.

amnn commented on crates/sui-package-resolver/src/lib.rs:
```diff
@@ -25,82 +18,18 @@ use move_core_types::{
-    #[error("Unexpected type: 'signer'.")]
-    UnexpectedSigner,
-}
+// TODO Move to ServiceConfig
```
This TODO belongs to the `PACKAGE_CACHE_SIZE` constant.

amnn reviewed (commented):
The use of the package cache looks correct and the overall extraction looks good, but it would be great if we could avoid the new `sui-package-resolver` crate from being a build graph bottleneck for all the different DB interfaces that implement its traits, so can we invert that? (i.e. have the consumer crates implement these interfaces?)

Additionally, seeing this crate used with another storage interface, and prompted by our discussion about the LruCache, I wonder whether it makes sense to get rid of that from within `PackageCache`, and instead push that into the specific implementation of `PackageStore` that the GraphQL service uses?

bmwill commented:
> Additionally, seeing this crate used with another storage interface, and prompted by our discussion about the LruCache, I wonder whether it makes sense to get rid of that from within PackageCache, and instead push that into the specific implementation of PackageStore that the GraphQL service uses?

That's an interesting idea, and i think it could be a bit more composable. I do think that the LruCache is probably useful outside of just the GraphQL service so maybe we structure things as a series of layers that implement package fetching.

- Cache Layer
- local storage layer (rocksdb)
- remote layer (fullnode or postgres)

And then you can mix and match all you like.

sadhansood commented:
Addressed comments:
* postgres and rocksdb dependencies are optional features now
* usage of lru cache is optional and rocksdb based package store completely bypasses the cache usage
* removed update() api from store api and rocksdb store handles update on its own (without going through the package store interface)

bmwill commented on crates/sui-package-resolver/Cargo.toml:
```diff
@@ -0,0 +1,43 @@
+tower.workspace = true
+
+[features]
+pg_integration = ["dep:diesel", "dep:sui-indexer"]
+rocksdb_integration = ["dep:rocksdb", "dep:tempfile", "dep:typed-store-derive", "dep:typed-store"]
```
I think per the conversation here we wanted to have these implementations live in separate crates or where they're being use. https://github.com/MystenLabs/sui/pull/14565#discussion_r1383550672

sadhansood replied:
Moved the `package_store` implementations to where they are being used

bmwill reviewed (approved):
There's a few small nits or tweaks that we could make on top of this, but the bulk of the changes that I was interested in happening are done so i'm happy to approve and unblock the analytics work.

amnn commented on crates/sui-package-resolver/src/lib.rs:
```diff
@@ -0,0 +1,1326 @@
+    };
+}
+
+as_ref_impl!(Arc<dyn PackageStore>);
```
neat!
